h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -720,15 +786,23 @@ public boolean converged(){
     return converged;
   }
 
-  protected double updateState(double [] beta,GLMGradientInfo ginfo){
-    _betaDiff = ArrayUtils.linfnorm(_beta == null?beta:ArrayUtils.subtract(_beta,beta),false);
+  protected double updateState(double [] beta, GLMGradientInfo ginfo) {
+    double[] oldBeta = changeArrayWActiveCols(_beta, true);
+    _betaDiff = ArrayUtils.linfnorm(_beta == null ? beta : ArrayUtils.subtract(oldBeta, beta), false);
     double objOld = objective();
-    if(_beta == null)_beta = beta.clone();
-    else System.arraycopy(beta,0,_beta,0,beta.length);
-    _ginfo = ginfo;
-    _likelihood = ginfo._likelihood;
-    return (_relImprovement = (objOld - objective())/Math.abs(objOld));
-  }
+    if (_beta == null) {
+      _beta = beta.clone();
+    } else {
+      if (_activeData._activeCols == null)
+        System.arraycopy(beta, 0, _beta, 0, beta.length);
+      else
+        System.arraycopy(changeArrayWActiveCols(beta, false), 0, _beta, 0, oldBeta.length);
+    }
+
+      _ginfo = ginfo;
+      _likelihood = ginfo._likelihood;
+      return (_relImprovement = (objOld - objective()) / Math.abs(objOld));
```

Comment:
Weird, in the second iteration (`_iter == 1`) of this failing test [1], we have lower `objOld` than `objective()` on this line. This leads to the test failure since without early stopping we stop prematurely (`relImprovement < 0 < eps`, line 782 in this file (`converged` method)).

Using absolute value of `_relImprovement` mitigates this problem but I think it might not be a proper solution (since there might be some change in behavior).

[1] `testdir_algos/glm/pyunit_PUBDEV_7611_early_stop_glm_multinomial.py`

Reply:
Tomas:  there are bugs I fixed that caused the coefficients to shift to the wrong place.  Hopefully this current fix will do the job.